test/parachain/tests/integration_test.rs

```diff
@@ -0,0 +1,399 @@
+}
+
+impl<'a> Drop for ChildHelper<'a> {
+	fn drop(&mut self) {
+		let name = self.name.clone();
+
+		self.terminate();
+
+		let mut stdout = String::new();
+		if let Some(reader) = self.child.stdout.as_mut() {
+			let _ = reader.read_to_string(&mut stdout);
+		}
+		eprintln!("process '{}' stdout:\n{}\n", name, stdout,);
+
+		let mut stderr = String::new();
+		if let Some(reader) = self.child.stderr.as_mut() {
+			let _ = reader.read_to_string(&mut stderr);
+		}
+		eprintln!("process '{}' stderr:\n{}\n", name, stderr,);
+	}
+}
+
+impl<'a> ChildHelper<'a> {
+	fn new(name: &str, child: &'a mut Child) -> ChildHelper<'a> {
+		ChildHelper {
+			name: name.to_string(),
+			child,
+		}
+	}
+
+	fn terminate(&mut self) {
+		match self.child.try_wait() {
+			Ok(Some(_)) => return,
+			Ok(None) => {}
+			Err(err) => {
+				eprintln!("could not wait for child process to finish: {}", err);
+				let _ = self.child.kill();
+				let _ = self.child.wait();
+				return;
+			}
+		}
+
+		#[cfg(unix)]
+		{
+			use nix::sys::signal::{kill, Signal::SIGTERM};
+			use nix::unistd::Pid;
+
+			kill(Pid::from_raw(self.child.id().try_into().unwrap()), SIGTERM).unwrap();
+
+			let mut tries = 30;
+
+			let success = loop {
+				tries -= 1;
+
+				match self.child.try_wait() {
+					Ok(Some(_)) => break true,
+					Ok(None) if tries == 0 => break false,
+					Ok(None) => thread::sleep(Duration::from_secs(1)),
+					Err(err) => {
+						eprintln!("could not wait for child process to finish: {}", err);
+						break false;
+					}
+				}
+			};
+
+			if !success {
+				let _ = self.child.kill();
+			}
+		}
+
+		#[cfg(not(unix))]
+		let _ = self.child.kill();
+
+		let _ = self.child.wait();
+	}
+}
+
+fn tcp_port_is_open<A: net::ToSocketAddrs>(address: A) -> bool {
+	net::TcpStream::connect(&address).is_ok()
+}
+
+async fn wait_for_tcp<A: net::ToSocketAddrs + std::fmt::Display>(address: A) {
+	loop {
+		match net::TcpStream::connect(&address) {
+			Ok(_) => break,
+			Err(err) => {
+				eprintln!("Waiting for {} to be up ({})...", address, err);
+				sleep(Duration::from_secs(2)).await;
+			}
+		}
+	}
+}
+
+#[async_std::test]
+#[ignore]
+async fn integration_test() {
+	assert!(
+		!tcp_port_is_open("127.0.0.1:9933"),
+		"tcp port is already open 127.0.0.1:9933, this test cannot be run",
+	);
+	assert!(
+		!tcp_port_is_open("127.0.0.1:9934"),
+		"tcp port is already open 127.0.0.1:9934, this test cannot be run",
+	);
+
+	let t1 = sleep(Duration::from_secs(60 * 10)).fuse();
+	let t2 = async {
+		// start alice
+		let polkadot_alice_dir = tempdir().unwrap();
+		let mut polkadot_alice = Command::new(cargo_bin("cumulus-test-parachain-collator"))
+			.stdout(Stdio::piped())
+			.stderr(Stdio::piped())
+			.args(POLKADOT_ARGS)
+			.arg("--base-path")
+			.arg(polkadot_alice_dir.path())
+			.arg("--alice")
+			.arg("--unsafe-rpc-expose")
+			.spawn()
+			.unwrap();
+		let polkadot_alice_helper = ChildHelper::new("alice", &mut polkadot_alice);
+		wait_for_tcp("127.0.0.1:9933").await;
```

Comment:
Test might run slightly faster if we start both Alice and Bob's nodes, then join on both `wait_for_tcp` calls.

Reply:
Really not worth the effort imo

Reply:
(also not really good for readability)

Reply:
Sure, agree it's not likely a huge savings.